File: java/src/test/java/challenges/aoc2024/day4/service/WordsearchServiceTest.java
package challenges.aoc2024.day4.service;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import challenges.aoc2024.day4.model.Wordsearch;

class WordsearchServiceTest {
    private WordsearchService wordsearchService;

    @BeforeEach
    void setUp() {
        wordsearchService = new WordsearchService();
    }

    @Test
    void testCountInstancesOfXmas() {
        // Given
        final List<String> wordsearchRows = List.of(
            "MMMSXXMASM",
            "MSAMXMSMSA",
            "AMXSXMAAMM",
            "MSAMASMSMX",
            "XMASAMXAMM",
            "XXAMMXXAMA",
            "SMSMSASXSS",
            "SAXAMASAAA",
            "MAMMMXMMMM",
            "MXMXAXMASX"
        );

        final Wordsearch wordsearch = Wordsearch.of(wordsearchRows);

        // When
        final Integer xmasCount = wordsearchService.countInstancesOfXmas(wordsearch);

        // Then
        assertEquals(18, xmasCount);
    }
}

File: scala/src/test/scala/challenges/aoc2024/day4/service/WordsearchServiceTest.scala
package challenges.aoc2024.day4.service

import challenges.aoc2024.day4.model.Wordsearch
import org.scalatest.BeforeAndAfter
import org.scalatest.funsuite.AnyFunSuite

class WordsearchServiceTest extends AnyFunSuite with BeforeAndAfter {

  var wordsearchService: WordsearchService = _

  before {
    wordsearchService = WordsearchService()
  }

  test("should count instances of xmas") {
    // Given
    val wordsearchRows: List[String] = List(
      "MMMSXXMASM",
      "MSAMXMSMSA",
      "AMXSXMAAMM",
      "MSAMASMSMX",
      "XMASAMXAMM",
      "XXAMMXXAMA",
      "SMSMSASXSS",
      "SAXAMASAAA",
      "MAMMMXMMMM",
      "MXMXAXMASX"
    )

    val wordsearch: Wordsearch = Wordsearch.of(wordsearchRows)

    // When
    val xmasCount: Int = wordsearchService.countInstancesOfXmas(wordsearch)

    // Then
    assert(18 == xmasCount)
  }

  test("should count instances of masx") {
    // Given
    val wordsearchRows: List[String] = List(
      "MMMSXXMASM",
      "MSAMXMSMSA",
      "AMXSXMAAMM",
      "MSAMASMSMX",
      "XMASAMXAMM",
      "XXAMMXXAMA",
      "SMSMSASXSS",
      "SAXAMASAAA",
      "MAMMMXMMMM",
      "MXMXAXMASX"
    )

    val wordsearch: Wordsearch = Wordsearch.of(wordsearchRows)

    // When
    val masxCount: Int = wordsearchService.countInstancesOfMasX(wordsearch)

    // Then
    assert(9 == masxCount)
  }
}
